lib/WasmReader/WasmByteCodeGenerator.cpp

```diff
@@ -8,56 +8,101 @@
+#define DebugPrintOp(op) if (PHASE_TRACE(Js::WasmReaderPhase, m_body)) { PrintOpName(op); }
 #else
 #define DebugPrintOp(op)
 #endif
 
 namespace Wasm
 {
-WasmBytecodeGenerator::WasmBytecodeGenerator(Js::ScriptContext * scriptContext, Js::Utf8SourceInfo * sourceInfo, byte* binaryBuffer, uint binaryBufferLength) :
-    m_scriptContext(scriptContext),
-    m_sourceInfo(sourceInfo),
-    m_alloc(_u("WasmBytecodeGen"), scriptContext->GetThreadContext()->GetPageAllocator(), Js::Throw::OutOfMemory),
-    m_reader(scriptContext->GetThreadContext()->GetPageAllocator(), binaryBuffer, binaryBufferLength),
-    m_f32RegSlots(nullptr),
-    m_f64RegSlots(nullptr),
-    m_i32RegSlots(nullptr),
-    m_currentFunc(nullptr),
-    m_evalStack(&m_alloc),
-    m_blockInfos(&m_alloc)
+#if DBG_DUMP
+    void
+        WasmBytecodeGenerator::PrintOpName(WasmOp op) const
+    {
+        switch (op)
+        {
+#define WASM_KEYWORD(token, name) \
+    case wn##token: \
+        Output::Print(_u(#token ## "\r\n")); \
+        break;
+#include "WasmKeywords.h"
+        }
+    }
+#endif
+
+/* static */
+Js::AsmJsRetType
+    WasmToAsmJs::GetAsmJsReturnType(WasmTypes::WasmType wasmType)
 {
-    m_writer.Create();
+    switch (wasmType)
+    {
+    case WasmTypes::F32:
+        return Js::AsmJsRetType::Float;
+    case WasmTypes::F64:
+        return Js::AsmJsRetType::Double;
+    case WasmTypes::I32:
+        return Js::AsmJsRetType::Signed;
+    case WasmTypes::Void:
+        return Js::AsmJsRetType::Void;
+    case WasmTypes::I64:
+        throw WasmCompilationException(_u("I64 support NYI"));
+    default:
+        throw WasmCompilationException(_u("Unknown return type %u"), wasmType);
+    }
+}
 
-    // TODO (michhol): try to make this more accurate?
-    const long astSize = 0;
-    m_writer.InitData(&m_alloc, astSize);
+/* static */
+Js::AsmJsVarType
+    WasmToAsmJs::GetAsmJsVarType(WasmTypes::WasmType wasmType)
+{
+    Js::AsmJsVarType asmType = Js::AsmJsVarType::Int;
+    switch (wasmType)
+    {
+    case WasmTypes::F32:
+        return Js::AsmJsVarType::Float;
+    case WasmTypes::F64:
+        return Js::AsmJsVarType::Double;
+    case WasmTypes::I32:
+        return Js::AsmJsVarType::Int;
+    case WasmTypes::I64:
+        throw WasmCompilationException(_u("I64 support NYI"));
+    default:
+        throw WasmCompilationException(_u("Unknown var type %u"), wasmType);
+    }
+}
+
+typedef bool(*SectionProcessFunc)(WasmModuleGenerator*);
+typedef void(*AfterSectionCallback)(WasmModuleGenerator*);
+
+WasmModuleGenerator::WasmModuleGenerator(Js::ScriptContext * scriptContext, Js::Utf8SourceInfo * sourceInfo, byte* binaryBuffer, uint binaryBufferLength) :
+    m_sourceInfo(sourceInfo),
+    m_scriptContext(scriptContext),
+    m_alloc(scriptContext->GetThreadContext()->GetThreadAlloc())
+{
+    m_reader = Anew(m_alloc, Binary::WasmBinaryReader, m_alloc, binaryBuffer, binaryBufferLength);
```

Comment:
also, since this data structure is living forever now, we may want to think about merging some of this with asm.js info (e.g. parameters, return types, etc.), which is now redundantly stored

---
In reply to: [71214068](https://github.com/Microsoft/ChakraCore/pull/1243#discussion_r71214068) [](ancestors = 71214068)